pkg/plugins/golang/v3/scaffolds/internal/templates/dockerignore.go

```diff
@@ -40,7 +40,10 @@ func (f *DockerIgnore) SetTemplateDefaults() error {
 
 const dockerignorefileTemplate = `# More info: https://docs.docker.com/engine/reference/builder/#dockerignore-file
 # Ignore all files which are not go type
-!**/*.go
-!**/*.mod
-!**/*.sum
+*
+!go.mod
+!go.sum
+!main.go
+!controllers/**/*.go
```

Comment:
We need to check if it will work when is multigroup when is not since it means that it can be:
https://github.com/kubernetes-sigs/kubebuilder/tree/master/testdata/project-v3-multigroup/controllers
OR
https://github.com/kubernetes-sigs/kubebuilder/tree/master/testdata/project-v3/controllers

Reply:
Good call.

I just tested it and it does not work. Further investigation lead me to this issue: https://github.com/moby/moby/issues/30018. Basically, we can't write a whitelist-style `.dockerignore` file.

Should we opt for a different strategy, like ignoring directories that are known to contain large files that are unnecessary when building a Docker image for the operator? The `.dockerignore` file would look like this:

```
# More info: https://docs.docker.com/engine/reference/builder/#dockerignore-file
# Ignore build and test binaries.
bin/
testbin/
```

Reply:
Hi @busser, 

If the previous suggestion works when is not multigroup then, see that we can:

- (option A) duplicate that to add the respective paths for multigroup by default as well
- (option B) we can change the file as we change  the dockerfile when we enable the multi-group support
- (option C) we can check a syntax that will ignore both paths.(I think it is the best solution)

Reply:
see : 

> Beyond Go’s filepath.Match rules, Docker also supports a special wildcard string ** that matches any number of directories (including zero). For example, **/*.go will exclude all files that end with .go that are found in all directories, including the root of the build context.

https://docs.docker.com/engine/reference/builder/#dockerignore-file

So, according to the docker documentation `!**/*.go` is correct which means that we are generating the file right.

Reply:
Hi @camilamacedo86,

The pattern we currently use, `!**/*.go` should work according to the Docker docs, but in fact doesn't. My understanding is that we are facing a bug in how Docker handles `.dockerignore` logic. This issue (https://github.com/moby/moby/issues/30018) reports a three year-old bug in Docker: when a pattern starts with `!`, wildcard strings like `*` and `**` don't work for directories.

I don't believe this will ever be fixed. The Docker team's stated long-term goal is for `.dockerignore` to have the same behaviour as `.gitignore`. I found this section in the [`.gitignore` specification](https://git-scm.com/docs/gitignore#_pattern_format):

> It is not possible to re-include a file if a parent directory of that file is excluded. Git doesn’t list excluded directories for performance reasons, so any patterns on contained files have no effect, no matter where they are defined.

So I guess that we won't ever be able to exclude the entire working directory except for GO files... 😞

Reply:
In its current state, the `.dockerignore` file scaffolded by Kubebuilder ignores nothing and includes the entire directory in the Docker build context. Most of the files are just code and are relatively small. However, the Kubebuilder workflow adds large binaries to the `bin/` and `testbin/` directories. I would recommend ignoring these directories by default and letting the user add more paths to their `.dockerignore` based on their specific needs.

What do you think?

Reply:
Would not the following configuration solve the need for multi-group and not multi-group layout projects?

```
!go.mod
!go.sum
!main.go
!controllers/**/*.go
!api/**/*.go
!controllers/*.go
!api/*.go
```

Reply:
I will close this one to make it simpler other follow up.

Reply:
What about the `config` directory? We should probably add a marker so that plugins can add their own exceptions, and that way the config plugin would be the one adding that line.

Reply:
@camilamacedo86 I've update the pull request to ignore the build and test binaries, as decided above 👍

Reply:
@Adirio How would you go about doing this? I don't mind doing the implementation myself, but would need some guidance 🙂